garage/tf/models/gaussian_mlp_model.py

```diff
@@ -0,0 +1,206 @@
+                 min_std=1e-6,
+                 max_std=None,
+                 std_hidden_nonlinearity=tf.nn.tanh,
+                 std_output_nonlinearity=None,
+                 std_parameterization='exp',
+                 hidden_nonlinearity=tf.nn.tanh,
+                 output_nonlinearity=None):
+        # Network parameters
+        super().__init__(name)
+        self._hidden_sizes = hidden_sizes
+        self._output_dim = output_dim
+        self._learn_std = learn_std
+        self._init_std = init_std
+        self._adaptive_std = adaptive_std
+        self._std_share_network = std_share_network
+        self._std_hidden_sizes = std_hidden_sizes
+        self._min_std = min_std
+        self._max_std = max_std
+        self._std_hidden_nonlinearity = std_hidden_nonlinearity
+        self._std_output_nonlinearity = std_output_nonlinearity
+        self._std_parameterization = std_parameterization
+        self._hidden_nonlinearity = hidden_nonlinearity
+        self._output_nonlinearity = output_nonlinearity
+
+        # Tranform std arguments to parameterized space
+        self._init_std_param = None
+        self._min_std_param = None
+        self._max_std_param = None
+        if self._std_parameterization == 'exp':
+            self._init_std_param = np.log(init_std)
+            if min_std:
+                self._min_std_param = np.log(min_std)
+            if max_std:
+                self._max_std_param = np.log(max_std)
+        elif self._std_parameterization == 'softplus':
+            self._init_std_param = np.log(np.exp(init_std) - 1)
+            if min_std:
+                self._min_std_param = np.log(np.exp(min_std) - 1)
+            if max_std:
+                self._max_std_param = np.log(np.exp(max_std) - 1)
+        else:
+            raise NotImplementedError
+
+    @property
+    def init_spec(self):
+        """
+        Model specification.
+
+        Positional and named arguments (args and kwargs) which, when passed
+        to `__init__`, will construct this model.
+        """
+        return (), {
+            'name': self._name,
+            'output_dim': self._output_dim,
+            'hidden_sizes': self._hidden_sizes,
+            'learn_std': self._learn_std,
+            'init_std': self._init_std,
+            'adaptive_std': self._adaptive_std,
+            'std_share_network': self._std_share_network,
+            'std_hidden_sizes': self._std_hidden_sizes,
+            'min_std': self._min_std,
+            'max_std': self._max_std,
+            'std_hidden_nonlinearity': self._std_hidden_nonlinearity,
+            'std_output_nonlinearity': self._std_output_nonlinearity,
+            'hidden_nonlinearity': self._hidden_nonlinearity,
+            'output_nonlinearity': self._output_nonlinearity,
+            'std_parameterization': self._std_parameterization,
+        }
+
+    def _build(self, *inputs):
+        action_dim = self._output_dim
+        state_input = inputs[0]
```

Comment:
What could inputs[1] be?

Reply:
inputs must be a tuple because there could be more than one.

i think this object should warn when there are too many inputs to unpack. perhaps we can make that a feature of the class `Model`

Reply:
e.g. Input can contain state and actions.

Reply:
Thanks for the explanation. I agree inputs should be a tuple. But I think using `state_input=inputs[0]` is weird. i.e. If this is a multi-goal env, the inputs would be observation plus goals, then the state_input should also be observation plus goals. If this is a Q function model, then state_input should be observation plus action.

Reply:
This is a difficult one to get right.

I chose `tuple` for the interface because it's universal and can use namedtuples if you want. However, using namedtuples for input would require that the model implementer provide you access to the type (e.g. as a class method). That's kind of ugly.

```python
class MyModel(TfModel):
    
    @classmethod
    def input_type(cls):
        return namedtuple('MyModelInputs', 'state', 'task')

    def _build(self, inputs):
        state = inputs.state

# ...
m = MyModel()
inputs = MyModel.input_type(state=my_state_var, task=my_task_var)  # this is a huge pain
out1, out2 = m.build(*inputs)

# also valid
m = MyModel()
inputs = MyModel()
out1, out2 = m.build(my_state_var, my_task_var)  # slight ambiguous at input, nice unpacking at the output
```

What if we make inputs/outputs a dict? (e.g. like kwargs).

That is a slightly more complicated interface than tuples but at least provides an opportunity for validation. 

I don't think it quite gets the dot-notation @CatherineSue wants, but it's close.

Unfortunately it makes unpacking really messy.

```python
class MyModel(TfModel)
    def build(self, **inputs):
        state_input = inputs['state']  # KeyError if it's missing
        # ...
        return dict(out1=out1, out2=out2)

# ...
m = MyModel()
outs = m.build(state=my_state_var, task=my_task_var)  # you can't unpack a dict because it's unordered
out1 = outs['out1']  # eww
out2 = outs['out2']
```

I think it's important to remember that kwargs can also be treated as positional arguments, which provides the solution here IMO

```python
class MyModel(TfModel):
    
    def _build(self, state=None, task=None):
        assert state, "You must provide a state input tensor"
        assert task, "You must provide a task input tensor"

        # ...
        return out1, out2
        
# ...
m = MyModel()
out1, out2 = m.build(state=my_state_var, task=my_task_var)

# also totally valid
m = MyModel()
out1, out2 = m.build(my_state_var, my_task_var)

# AssertionError
m = MyModel()
out1, out2 = m.build(my_state_var)

# AssertionError
m = MyModel()
out1, out2 = m.build(state=my_state_va)
```

Reply:
Then the interface would have `def _build(self)`? We would have to specify very carefully in docstring that inputs are expected, but the inputs depend on the Model. It can be `state`, or both `state` and `action`.

Reply:
or it will still be `def _build(self, *inputs)` and we will pass actual kwargs for each models

Reply:
No the interface would be unchanged.

```python
class BaseModel(abc.ABC):

    def build(self, *inputs):
        pass


class Model(BaseModel, metaclass=abc.ABCMeta):
    # ...


class TfModel(Model, metaclass=abc.ABCMeta):

    def build(self, *inputs, name=Nome):
        # ...

    def _build(self, *inputs, name=None):
        pass

class MyModel(TfModel):

    def _build(self, state, task, name=None):
        # ...
        return dist.sample(), mean, std



# Works fine
m = MyModel()
a, mean, std = m.build(my_state_var, my_task_var)

# Also works fine
m = MyModel()
a, mean, std = m.build(state=my_state_var, task=my_task_var, name='Foo')

# Also works fine
m = MyModel()
a, mean, std = m.build(state=my_state_var, task=my_task_var, 'Foo')

# Also works fine
m = MyModel()
a, mean, std = m.build(task=my_task_var, state=my_state_var, 'Foo')

# Error
m = MyModel()
a, mean, std = m.build(my_state_var, name='Foo')

# Error
m = MyModel()
a, mean, std = m.build(task=my_task_var)

# etc.
```

Reply:
I agree with @ryanjulian 's proposal. Another question maybe related to the outputs unpack here: what's the purpose of these fileds [here](https://github.com/rlworkgroup/garage/pull/512/files/96e6512f704cb687464f55823b68c5306a74d760#diff-f27d474d4aed442478e7e44fdc2de671R201)? I don't see where it's explicitly used.

Reply:
I tried to make the outputs to be attributes of the `Model`.

Reply:
the user should only be able to access the outputs of a model through the return value of `build` and through the `outputs` property of a net.

We can try it this way first. If it turns out that it makes for a lot of boilerplate code, we can come back and try something different with more magic.

```python
m = MyModel()
foo = m.build(bar)
assert m.output == foo  # AttributeError
assert m.networks['default'].output == foo
baz = m.build(bap, name='alice')
assert m.output == baz  # AttributeError
assert m.networks['alice'] == baz
```